src/odb/src/lefin/MetalWidthViaMapParser.cpp

```diff
@@ -0,0 +1,115 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <functional>
+#include <iostream>
+#include <string>
+
+#include "boostParser.h"
+#include "db.h"
+#include "lefLayerPropParser.h"
+#include "lefin.h"
+
+using namespace odb;
+void MetalWidthViaMapParser::addEntry(
+    boost::fusion::vector<std::string,
+                          double,
+                          double,
+                          boost::optional<double>,
+                          boost::optional<double>,
+                          std::string>& params)
+{
+  via_map = dbMetalWidthViaMap::create(tech_);
+  auto layer_name = at_c<0>(params);
+  if (at_c<3>(params).is_initialized()) {
+    via_map->setBelowLayerWidthLow(lefin_->dbdist(at_c<1>(params)));
+    via_map->setBelowLayerWidthHigh(lefin_->dbdist(at_c<2>(params)));
+    via_map->setAboveLayerWidthLow(lefin_->dbdist(at_c<3>(params).value()));
+    via_map->setAboveLayerWidthHigh(lefin_->dbdist(at_c<4>(params).value()));
+  } else {
+    via_map->setBelowLayerWidthLow(lefin_->dbdist(at_c<1>(params)));
+    via_map->setBelowLayerWidthHigh(lefin_->dbdist(at_c<1>(params)));
+    via_map->setAboveLayerWidthLow(lefin_->dbdist(at_c<2>(params)));
+    via_map->setAboveLayerWidthHigh(lefin_->dbdist(at_c<2>(params)));
+  }
+  via_map->setViaName(at_c<5>(params));
+  via_map->setViaCutClass(cut_class_);
+  incomplete_props_.push_back({via_map, layer_name});
+}
+
+void MetalWidthViaMapParser::setCutClass()
+{
+  cut_class_ = true;
+}
+
+void MetalWidthViaMapParser::setPGVia()
+{
+  via_map->setPgVia(true);
+}
+
+void MetalWidthViaMapParser::parse(std::string s)
+{
+  std::vector<std::string> rules;
+  boost::split(rules, s, boost::is_any_of(";"));
+  for (auto rule : rules) {
+    boost::algorithm::trim(rule);
+    if (rule.empty())
+      continue;
+    rule += " ; ";
+    if (!parseSubRule(rule))
+      lefin_->warning(299,
+                      "parse mismatch in propery LEF58_METALWIDTHVIAMAP"
+                      ":\"{}\"",
+                      rule);
+  }
+}
+
+bool MetalWidthViaMapParser::parseSubRule(std::string s)
+{
+  cut_class_ = false;
+  qi::rule<std::string::iterator, std::string(), ascii::space_type> _string;
```

Comment:
Don't use leading (or trailing) underscore on local variable names.

Reply:
I am trying to match the boost spirit terminals (e.x. double_ int_). Is that ok? if not then it needs to be changed in all lefin.

Reply:
@maliberty I am gonna merge and we can address this in a separate PR if it's an issue.

Reply:
It can be fixed in a separate PR.  I see in some cases you used this style and other you used appended "Rule" to the name which I prefer (eg lefTechLayerCornerSpacingParser.cpp)